firmware/views.py

```diff
@@ -0,0 +1,178 @@
+# coding: utf8
+""" Routes for app paths and functions combining multiple modules """
+
+
+import sqlalchemy.orm
+from flask import request, session, redirect, url_for, render_template, flash
+from firmware import app
+from firmware import repository
+from firmware.models import Company, User, Review
+from firmware.uploaders import upload_file
+from firmware.forms import AddUser, AddCompany, LogIn, AddReview
+from firmware.decorators import login_required
+from firmware.authorization import authorize_edit_company
+
+
+
+def get_details(company_id):
+    """get details about a company via it's id"""
+    try:
+        company = repository.get_company(company_id)
+    except sqlalchemy.orm.exc.NoResultFound:
+        errors = dict()
+        errors['company-not-found'] = "Company not found!"
+        return render_template('404.html', error_messages=errors)
+    category = company.category.domain
+    reviews = repository.get_reviews(company_id)
+    added_by_user = company.added_by_user.username
+    return company, category, reviews, added_by_user
+
+
+@app.route('/home')
+def home():
+    """ function called when home path (url) is accessed"""
+    category = request.args.get('category', 'all categories')
+    if category == 'all categories':
+        companies = repository.get_companies()
+    else:
+        companies = repository.get_filtered_companies(category)
+    return render_template('index.html', companies=companies,
+                           categories=repository.get_categories(),
+                           current_category=category)
+
+
+@app.route('/404-page-not-found')
+def not_found(errors):
+    """ function to render page for error 404 """
+    return render_template('404.html', error_messages=errors)
+
+
+@app.route('/details/<company_id>', methods=['GET', 'POST'])
+def details(company_id):
+    """ function to show details about a company via it's id """
+    errors = dict()
+    try:
+        company, category, reviews, added_by_user = get_details(company_id)
+        add_review_form = AddReview(request.form)
+    except ValueError:
+        errors['Company-not-found'] = "Company not found!"
+        return render_template('404.html', error_messages=errors)
+    if request.method == 'POST' and add_review_form.validate():
```

Comment:
You can move the code for adding a review in another action so that you can use the decorator which checks the user.